### PR TITLE
[SPARK-12719][SQL] [WIP] SQL generation support for generators, including UDTF

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?

This is a alternate way to convert SQL from analyzed logical plans containing Generate operator.
In this PR , generators in projection list are expressed as LATERAL VIEW.

Sample Plan :

```
GlobalLimit 3
+- LocalLimit 3
   +- Project [gencol2#204]
      +- Generate explode(gencol1#203), true, false, Some(gentab2), [gencol2#204]
         +- Generate explode(array(array(1, 2, 3))), true, false, Some(gentab1), [gencol1#203]
            +- MetastoreRelation default, t4, None
```

Generated Query:

```
SELECT `gentab2`.`gencol2` FROM `default`.`t4` LATERAL VIEW explode(array(array(1, 2, 3))) `gentab1` AS `gencol1` LATERAL VIEW explode(`gentab1`.`gencol1`) `gentab2` AS `gencol2` LIMIT 3
```
## How was this patch tested?

Tests added to LogicalPlanToSQLSuite

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by dilipbiswal** [comment]
@rxin @cloud-fan @liancheng @gatorsmile  Hi Reynold and Wenchen, 
I have made change to generate LATERAL view for all generators specified in projection list as well as in from clause. I am still testing it and adding more tests. Please let me know what you think..

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
hi @dilipbiswal , thanks for working on this!  Some high-level questions:

1. Can any `Generate` operator  be converted to a LATERAL VIEW? Does the plan tree have to match some kind of pattern?
2. Can any `Generate` operator  be converted to a "SELECT ...udtf..." format?

Generally speaking, we should use one unified format for `Generator`'s SQL string if possible, whatever the original SQL string is.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 1]
```diff
@@ -297,6 +303,60 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
...
+      outerClause,
+      plan.generator.sql,
+      quoteIdentifier(generatorAlias),
+      "AS",
+      columnAliases
+    )
+  }
+
+  private def generateToSQL(plan: Project): String = {
+    // assert if child is a generate or not.
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Please add an assert here. 
```
assert(join == true || plan.projectList == 1)
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 2]
```diff
@@ -94,6 +94,12 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
     case Distinct(p: Project) =>
       projectToSQL(p, isDistinct = true)
 
+    case p@ Project(_, g: Generate) =>
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Nit: added a space after `p`

**Comment by gatorsmile** [comment]
LGTM

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 3]
```diff
@@ -297,6 +303,60 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
     )
   }
 
+  /* This function handles the SQL generation when generators.
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: style issue: please use `/**`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala [hunk 4]
```diff
@@ -445,4 +461,98 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+    checkHiveQl(
+      """
+        |SELECT subq.key
+        |FROM
+        |(SELECT key, value from t1 LATERAL VIEW explode(value) gentab AS gencol) subq
+      """.stripMargin
+    )
+  }
+
+  test("SQL generation for UDTF") {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Please add more UDTF cases, as what @cloud-fan suggested.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala [hunk 5]
```diff
@@ -445,4 +461,98 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+
+    // non-generated columns in projection list
+    checkHiveQl("SELECT key as c1, explode(array(1,2,3)) as c2, value as c3 FROM t1")
+  }
+
+  test("SQL generation for json_tuple as generator") {
+    checkHiveQl("SELECT key, json_tuple(jstring, 'f1', 'f2', 'f3', 'f4', 'f5') FROM parquet_t3")
+  }
+
+  test("SQL generation for lateral views") {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Please add another test case for joining the lateral view with a regular table. Thanks! 

```LATERAL VIEW ......,  t1 as table1```

**Comment by dilipbiswal** [comment]
@cloud-fan Thank you.
1. Can any Generate operator be converted to a LATERAL VIEW? Does the plan tree have to match some kind of pattern?

Yes, based on the tests i have tried. I am testing this more to see if i uncover any issue. Now i am converting
```SQL
select explode(array(1,2,3)) from t1
``` 
to
```SQL
select tablequalifier.colname from t1 LATERAL VIEW explode(array(1,2,3)) tablequalifier as colname
``` 
2. Can any Generate operator be converted to a "SELECT ...udtf..." format?

No. As we have a restriction of allowing one generate in a projection list. Thats why we try to
use the LATERAL VIEW clause.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala (hunk 2 above)
Comment (dilipbiswal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Will do. Thanks.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala (hunk 3 above)
Comment (dilipbiswal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Will do. Thanks.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala (hunk 4 above)
Comment (dilipbiswal) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: OK.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala (hunk 5 above)
Comment (dilipbiswal) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: OK.

**Comment by cloud-fan** [comment]
But a `Generate` operator only has one `generator`, so I think we can convert any kind of `Generate` to SELECT?

**Comment by dilipbiswal** [comment]
@cloud-fan Yeah.. that should be possible, You are thinking to represent each Generate as a sub select clause ?

**Comment by cloud-fan** [comment]
I'm not sure which one is better, it depends on which one is easier to understand and implement. I'm ok to generate verbose SQL string, but the generator itself should be as simple and robust as possible.

**Comment by dilipbiswal** [comment]
@cloud-fan Thanks !! Actually i remember exploring this option. There is a generate syntax for outer lateral view like following -
```SQL
SELECT * FROM src LATERAL VIEW OUTER explode(array()) C AS a limit 10;
```
I thought if we converted this generate operator as  a sub select where the generator is in projection 
list , we will loose the outer semantics. Thats why i felt safer to generate as a LATERAL VIEW
clause.

**Comment by cloud-fan** [comment]
ah makes sense, then we should always generate a LATERAL VIEW format SQL.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 6]
```diff
@@ -297,6 +303,65 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
...
+   * sample plan :
+   *   +- Project [mycol2#192]
+   *     +- Generate explode(myCol#191), true, false, Some(mytable2), [mycol2#192]
+   *       +- Generate explode(array(array(1, 2, 3))), true, false, Some(mytable), [mycol#191]
+   *         +- MetastoreRelation default, src, None
+   *
+   */
+  private def generateToSQL(plan: Generate): String = {
+    val columnAliases = plan.generatorOutput.map(a => quoteIdentifier(a.name)).mkString(",")
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: you can just do `plan.generatorOutput.map(_.sql).mkString(", ")`

Comment (dilipbiswal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Sure. Thx

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 7]
```diff
@@ -94,6 +94,12 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
     case Distinct(p: Project) =>
       projectToSQL(p, isDistinct = true)
 
+    case p@ Project(_, g: Generate) =>
+      generateToSQL(p)
+
+    case g: Generate =>
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: not quite sure, but can we convert this to `Project(g.outpur, g)` and share the same code path with the above case?

Comment (dilipbiswal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: @cloud-fan Sorry Wenchen for getting back to you a bit late. Actually the two cases do different things. First one builds the top level SELECT clause and the second one builds the LATERAL VIEW clause. Do you have any suggestion for a better name for these two functions ? :-)

**Comment by dilipbiswal** [comment]
@cloud-fan Can you please help trigger a retest ? This does not seem related to my changes.

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
Hi @dilipbiswal , sorry that recently I made a major refactor to `SQLBuilder`, and it changed some assumptions to the `Generate` SQL generation, so I took over your PR and send a new one with some simplification: #11696, do you mind taking a look?

**Comment by dilipbiswal** [comment]
Closing this in favor of 
[#11696] (http://github.com/apache/spark/pull/11696)

**Pull request closed without merging**